pkg/apis/v1beta1/taints.go

```diff
@@ -32,6 +34,15 @@ var (
 		Effect: v1.TaintEffectNoSchedule,
 		Value:  DisruptingNoScheduleTaintValue,
 	}
+
+	// DisruptionNonGracefulShutdown is used by the deprovisioning controller to forcefully
+	// shut down a node. This does not respect graceful termination of any pods on the node.
+	// https://kubernetes.io/docs/concepts/architecture/nodes/#non-graceful-node-shutdown
+	DisruptionNonGracefulShutdown = v1.Taint{
```

Comment:
I think that in the dominant happy path, the taint isn't necessary. I'm worried about the cases none of us have been able to consider. We could leave those exceptional cases up to the cluster admins to handle and identify workarounds for.

I think using the out-of-service does provide us some additional benefits in the form of PV handling.
> This taint triggers pods on the node to be forcefully deleted if there are no matching tolerations on the pods. Persistent volumes attached to the shutdown node will be detached, and new pods will be created successfully on a different running node.

It is an official K8s mechanism that is well-documented and GA, so it's very unlikely to shift much in the future.
https://kubernetes.io/blog/2023/08/16/kubernetes-1-28-non-graceful-node-shutdown-ga/

Reply:
> Before applying the out-of-service taint, you must verify that a node is already in shutdown or power-off state (not in the middle of restarting).

Are we doing that?

Reply:
Also see https://github.com/kubernetes-sigs/karpenter/pull/916#pullrequestreview-2133945824

Reply:
That's a fair point, though I suspect it's just the k8s maintainers being overly careful and not something that would ever become an issue since our goal is to demolish the node in question anyway. 

I'm fine with pulling out the taint logic since it'll simplify the implementation of this feature, as long as it still accomplishes everything it needs to. Let me verify that the desired behavior without the taint handles pods with infinite prestop hooks in particular.

I would have preferred for this concern about the core mechanism that was implemented in the first iteration of the PR back in December 2023 to have been raised earlier rather than at the 11th hour.

Reply:
Chatted more about this in [#karpenter-dev](https://kubernetes.slack.com/archives/C04JW2J5J5P/p1719094373267959).

I've verified that the feature is working as intended without use of any taint being applied to the node.